db/flush_scheduler.cc

```diff
@@ -15,11 +15,9 @@ namespace rocksdb {
 
 void FlushScheduler::ScheduleFlush(ColumnFamilyData* cfd) {
 #ifndef NDEBUG
-  {
-    std::lock_guard<std::mutex> lock(checking_mutex_);
-    assert(checking_set_.count(cfd) == 0);
-    checking_set_.insert(cfd);
-  }
+  std::lock_guard<std::mutex> lock(checking_mutex_);
```

Comment:
I wonder why this lock is brought to outside the block? The result would be that in debug mode the entire :: ScheduleFlush is protected by a lock and in release mode, it is not.

Reply:
I don't recall the exact reason, but from the patch it seems to prevent race condition with `TakeNextColumnFamily` or `Empty`. The lock should not be used in release mode.

Reply:
It seems safe to reduce the scope, though.

Reply:
Thanks @yiwu-arbug for chiming in. We should use the same concurrency mechanism in debug mode as in the release mode so if there is a concurrency issue it would show up in our tsan tests.

Reply:
The lock here is to guard the `checking_set_` which is used in debug mode to verify correctness. Otherwise `FlushScheduler` is lock free.

Reply:
Of cause it would be great if we can remove the `checking_set_` entirely and come up with another way for verification without interfering the concurrency mechanism.

Reply:
I am confused. Why cannot we simply reduce the scope of the lock back to `checking_set_`. Why the entire function has to be protected by the lock?

Reply:
I left too many comments and probably confuse you. I mean, reduce the scope seems right :)